arch/xtensa/src/esp32s3/esp32s3_start.c

```diff
@@ -447,55 +488,155 @@ static inline uint32_t calc_mmu_pages(uint32_t size, uint32_t vaddr)
+        {
+          app_irom_start = offset + sizeof(esp_image_segment_header_t);
+          app_irom_start_aligned = app_irom_start & MMU_FLASH_MASK;
+          rom_segments++;
+        }
+
+      if (IS_SRAM(segment_hdr.load_addr))
+        {
+          ram_segments++;
+        }
+
+      offset += sizeof(esp_image_segment_header_t) + segment_hdr.data_len;
+      if (ram_segments == image_header.segment_count && !padding_checksum)
+        {
+          offset += (CHECKSUM_ALIGN - 1) - (offset % CHECKSUM_ALIGN) + 1;
+          padding_checksum = true;
+        }
+    }
 
-  for (size_t i = 0; i < FLASH_MMU_TABLE_SIZE; i++)
+  if (segments == 0 || segments == 16)
     {
-      FLASH_MMU_TABLE[i] = MMU_TABLE_INVALID_VAL;
+      ets_printf("Error parsing segments\n");
     }
 
-  drom_lma_aligned = app_drom_lma & MMU_FLASH_MASK;
-  drom_vma_aligned = app_drom_vma & MMU_FLASH_MASK;
-  drom_page_count = calc_mmu_pages(app_drom_size, app_drom_vma);
-  rc = cache_dbus_mmu_set(MMU_ACCESS_FLASH, drom_vma_aligned,
-                          drom_lma_aligned, 64, drom_page_count, 0);
+  ets_printf("total segments stored %d\n", segments - 1);
+#endif
+
+  cache_hal_disable(CACHE_TYPE_ALL);
+
+  /* Clear the MMU entries that are already set up,
+   * so the new app only has the mappings it creates.
+   */
 
-  irom_lma_aligned = app_irom_lma & MMU_FLASH_MASK;
-  irom_vma_aligned = app_irom_vma & MMU_FLASH_MASK;
-  irom_page_count = calc_mmu_pages(app_irom_size, app_irom_vma);
-  rc = cache_ibus_mmu_set(MMU_ACCESS_FLASH, irom_vma_aligned,
-                          irom_lma_aligned, 64, irom_page_count, 0);
+  mmu_hal_unmap_all();
 
-  regval  = getreg32(EXTMEM_DCACHE_CTRL1_REG);
-  regval &= EXTMEM_DCACHE_SHUT_CORE0_BUS;
-  putreg32(regval, EXTMEM_DCACHE_CTRL1_REG);
+  mmu_hal_map_region(0, MMU_TARGET_FLASH0,
+                     app_drom_vaddr_aligned, app_drom_start_aligned,
+                     app_drom_size, &actual_mapped_len);
 
-  cache_resume_dcache(autoload);
+  mmu_hal_map_region(0, MMU_TARGET_FLASH0,
+                     app_irom_vaddr_aligned, app_irom_start_aligned,
+                     app_irom_size, &actual_mapped_len);
+
+  /* ------------------Enable corresponding buses--------------------- */
+
+  cache_bus_mask_t bus_mask = cache_ll_l1_get_bus(0, app_drom_vaddr_aligned,
+                                                  app_drom_size);
+  cache_ll_l1_enable_bus(0, bus_mask);
+  bus_mask = cache_ll_l1_get_bus(0, app_irom_vaddr_aligned, app_irom_size);
+  cache_ll_l1_enable_bus(0, bus_mask);
+#if CONFIG_ESPRESSIF_NUM_CPUS > 1
+  bus_mask = cache_ll_l1_get_bus(1, app_drom_vaddr_aligned, app_drom_size);
+  cache_ll_l1_enable_bus(1, bus_mask);
+  bus_mask = cache_ll_l1_get_bus(1, app_irom_vaddr_aligned, app_irom_size);
+  cache_ll_l1_enable_bus(1, bus_mask);
+#endif
+
+  /* ------------------Enable Cache----------------------------------- */
+
+  cache_hal_enable(CACHE_TYPE_ALL);
```

Comment:
can these apis (cache_hal_enable/cache_hal_disable) be used w/o calling cache_hal_init?

Reply:
@almir-okato

Reply:
Hi @yamt , the `cache_hal_init` is called from `bootloader_init`. Same routine is called from MCUboot bootloader.

Reply:
> Hi @yamt , the `cache_hal_init` is called from `bootloader_init`. Same routine is called from MCUboot bootloader.

this one? https://github.com/espressif/esp-hal-3rdparty/blob/7c4fae7ebc5342b19cab03511bfa277176bba377/components/bootloader_support/src/esp32s3/bootloader_esp32s3.c#L214

Reply:
Yes

Reply:
ok. thank you.